ledger/ledgercore/statedelta.go

```diff
@@ -191,29 +195,87 @@ type AccountDeltas struct {
 	assetResourcesCache map[AccountAsset]int
 }
 
-// MakeStateDelta creates a new instance of StateDelta.
+// MakeStateDelta creates a new instance of StateDelta
 // hint is amount of transactions for evaluation, 2 * hint is for sender and receiver balance records.
 // This does not play well for AssetConfig and ApplicationCall transactions on scale
-func MakeStateDelta(hdr *bookkeeping.BlockHeader, prevTimestamp int64, hint int, stateProofNext basics.Round) StateDelta {
-	return StateDelta{
-		Accts: MakeAccountDeltas(hint),
-		Txids: make(map[transactions.Txid]IncludedTransactions, hint),
-		// asset or application creation are considered as rare events so do not pre-allocate space for them
-		Hdr:                      hdr,
-		StateProofNext:           stateProofNext,
-		PrevTimestamp:            prevTimestamp,
-		initialTransactionsCount: hint,
+func MakeStateDelta(hdr *bookkeeping.BlockHeader, prevTimestamp int64, hint int, stateProofNext basics.Round) (sd StateDelta) {
+	sd = StateDelta{}
+	sd.PopulateStateDelta(hdr, prevTimestamp, hint, stateProofNext)
+	return
+}
+
+// PopulateStateDelta populates an existing StateDelta struct.
+// Used as a helper for MakeStateDelta as well as for re-using already allocated structs from sync.Pool
+func (sd *StateDelta) PopulateStateDelta(hdr *bookkeeping.BlockHeader, prevTimestamp int64, hint int, stateProofNext basics.Round) {
+	if sd.Txids == nil {
+		sd.Txids = make(map[transactions.Txid]IncludedTransactions, hint)
+	}
+	if sd.Accts.isEmpty() {
+		sd.Accts = MakeAccountDeltas(hint)
 	}
+	sd.Hdr = hdr
+	sd.StateProofNext = stateProofNext
+	sd.PrevTimestamp = prevTimestamp
+	sd.initialTransactionsCount = hint
 }
 
 // MakeAccountDeltas creates account delta
+// if adding new fields make sure to add them to the .reset() and .isEmpty() methods
 func MakeAccountDeltas(hint int) AccountDeltas {
 	return AccountDeltas{
 		Accts:      make([]BalanceRecord, 0, hint*2),
 		acctsCache: make(map[basics.Address]int, hint*2),
 	}
 }
 
+// Reset resets the StateDelta for re-use with sync.Pool
+func (sd *StateDelta) Reset() {
+	sd.Accts.reset()
+	for txid := range sd.Txids {
+		delete(sd.Txids, txid)
+	}
+	for txLease := range sd.Txleases {
+		delete(sd.Txleases, txLease)
+	}
+	for creatableIndex := range sd.Creatables {
+		delete(sd.Creatables, creatableIndex)
+	}
+	for key := range sd.KvMods {
+		delete(sd.KvMods, key)
+	}
+	sd.Totals = AccountTotals{}
+
+	// these fields are going to be populated on next use but resetting them anyway for safety.
+	sd.Hdr = nil
+	sd.StateProofNext = basics.Round(0)
+	sd.PrevTimestamp = 0
+	sd.initialTransactionsCount = 0
+}
+
+// reset clears out allocated slices from AccountDeltas struct for reuse with sync.Pool
+func (ad *AccountDeltas) reset() {
+	// reset the slices
+	ad.Accts = ad.Accts[:0]
+	ad.AppResources = ad.AppResources[:0]
+	ad.AssetResources = ad.AssetResources[:0]
+
+	// reset the maps
+	for address := range ad.acctsCache {
```

Comment:
please put something into these maps to have loops actually executed + have a full func coverage